### PR TITLE
Fix TypicalLogitsWarper tensor OOB indexing edge case

### PR DESCRIPTION
**Pull request by njhill**
This can be triggered fairly quickly with low precision e.g. `bfloat16` and `typical_p=0.99`.

@gante

### CODE REVIEW

**Comment by Rocketknight1** [comment]
I'm going to be honest - I don't fully understand the code here!

The array `last_ind` is created as:
`last_ind = (cumulative_probs < self.mass).sum(dim=1)`

This is the sum of a boolean array, which should be strictly nonnegative, because boolean arrays only contain 0 and 1 values. Therefore, I don't understand why the original line `last_ind[last_ind < 0] = 0` or the replacement using `torch.clamp_` are necessary - I don't see how you'd get negative values without an integer overflow, and if we're getting an integer overflow we should be using bigger integer dtypes, not fixing it with value clamping.

Do you know why this is necessary in the first place?

**Comment by njhill** [comment]
@Rocketknight1 good point, I hadn't considered whether the existing check was valid/redundant, I guess it's [always been there](https://github.com/huggingface/transformers/pull/15504).

I guess it would make more sense if the prior line was `last_ind = (cumulative_probs < self.mass).sum(dim=1) - 1`, perhaps that was originally intended but left out. I'll update this PR accordingly.

**Comment by njhill** [comment]
I guess this will actually mean a subtle change in behaviour, but I'm fairly sure it's what was originally intended. Not sure whether this is ok though w.r.t. transformers policies around this kind of thing...

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -480,8 +480,8 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
         cumulative_probs = sorted_logits.softmax(dim=-1).cumsum(dim=-1)
 
         # Remove tokens with cumulative mass above the threshold
-        last_ind = (cumulative_probs < self.mass).sum(dim=1)
-        last_ind[last_ind < 0] = 0
+        last_ind = (cumulative_probs < self.mass).sum(dim=1) - 1
+        last_ind.clamp_(0)
```
Comment (gante) [inline src/transformers/generation/logits_process.py:484]:
```suggestion
        last_ind.clamp_(min=0)
```
nit: let's make it more clear that it is a minimum clamp

**Review by gante — APPROVED** [review]
Thank you for the PR @njhill!

This makes sense -- the last valid index of a sorted vector with N valid values is N-1

Since this is a bug fix, we don't need a deprecation cycle :)

File: src/transformers/generation/logits_process.py (hunk 1 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:484]: @Rocketknight1 After the change, the `(cumulative_probs < self.mass).sum(dim=1)` can be 0, which added to `-1` turns into a negative value. The cumsum vector at index N already contains the value of the tensor at index N, so if the first member of `cumulative_probs` is larger than `self.mass` a negative value will come out of here :D 

Before the change, it seemed to be a redundant op, yes 😂

Comment (Rocketknight1) [reply src/transformers/generation/logits_process.py:484]: That makes sense, lol. Just making sure I wasn't missing something incredibly obvious!

**Comment by gante** [comment]
Thank you for the fix @njhill 💪
